include/drivers/pinctrl/stm32.h

```diff
@@ -0,0 +1,126 @@
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+/**
+ * @brief Utility macro to initialize pincfg field (F1 series).
+ *
+ * @param node_id Node identifier.
+ * @param state_prop State property name.
+ * @param idx State property entry index.
+ */
+#define Z_PINCTRL_STM32_PINCFG_INIT(node_id, state_prop, idx)		       \
+	(((STM32_NO_PULL *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_disable))     \
+	   << STM32_PUPD_SHIFT) |					       \
+	 ((STM32_PULL_UP *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_pull_up))     \
+	   << STM32_PUPD_SHIFT) |					       \
+	 ((STM32_PULL_DOWN *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_pull_down))   \
+	   << STM32_PUPD_SHIFT) |					       \
+	 ((STM32_PUSH_PULL *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, drive_push_pull))  \
+	   << STM32_CNF_OUT_0_SHIFT) |					       \
+	 ((STM32_OPEN_DRAIN *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, drive_open_drain)) \
+	   << STM32_CNF_OUT_0_SHIFT) |					       \
+	 (DT_ENUM_IDX(DT_PROP_BY_IDX(node_id, state_prop, idx), slew_rate)     \
+	   << STM32_MODE_OSPEED_SHIFT))
+#else
+/**
+ * @brief Utility macro to initialize pincfg field (non-F1 series).
+ *
+ * @param node_id Node identifier.
+ * @param state_prop State property name.
+ * @param idx State property entry index.
+ */
+#define Z_PINCTRL_STM32_PINCFG_INIT(node_id, state_prop, idx)		       \
+	(((STM32_NO_PULL *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_disable))     \
+	   << STM32_PUPDR_SHIFT) |					       \
+	 ((STM32_PULL_UP *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_pull_up))     \
+	   << STM32_PUPDR_SHIFT) |					       \
+	 ((STM32_PULL_DOWN *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_pull_down))   \
+	   << STM32_PUPDR_SHIFT) |					       \
+	 ((STM32_PUSH_PULL *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, drive_push_pull))  \
+	   << STM32_OTYPER_SHIFT) |					       \
+	 ((STM32_OPEN_DRAIN *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, drive_open_drain)) \
+	   << STM32_OTYPER_SHIFT) |					       \
+	 (DT_ENUM_IDX(DT_PROP_BY_IDX(node_id, state_prop, idx), slew_rate)     \
+	   << STM32_OSPEEDR_SHIFT))
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
+
+/**
+ * @brief Utility macro to initialize each pin.
+ *
+ * @param node_id Node identifier.
+ * @param state_prop State property name.
+ * @param idx State property entry index.
+ */
+#define Z_PINCTRL_STATE_PIN_INIT(node_id, state_prop, idx)		       \
+	{ .pinmux = DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, pinmux),  \
+	  .pincfg = Z_PINCTRL_STM32_PINCFG_INIT(node_id, state_prop, idx) },
+
+/**
+ * @brief Utility macro to initialize state pins contained in a given property.
+ *
+ * @param node_id Node identifier.
+ * @param prop Property name describing state pins.
```

Comment:
Add one level of indirection to the macro's make's this a lot more readable:

```
#define Z_PINCTRL_STM32_PINCFG_INIT(node)                      \
        (((STM32_NO_PULL * DT_PROP(node, bias_disable)) << STM32_PUPD_SHIFT) |           \
         ((STM32_PULL_UP * DT_PROP(node, bias_pull_up)) << STM32_PUPD_SHIFT) |           \
         ((STM32_PULL_DOWN * DT_PROP(node, bias_pull_down)) << STM32_PUPD_SHIFT) |       \
         ((STM32_PUSH_PULL * DT_PROP(node, drive_push_pull)) << STM32_CNF_OUT_0_SHIFT) |  \
         ((STM32_OPEN_DRAIN * DT_PROP(node, drive_open_drain)) << STM32_CNF_OUT_0_SHIFT) |   \
         (DT_ENUM_IDX(node, slew_rate) << STM32_MODE_OSPEED_SHIFT))
#else
        ...
#endif /* CONFIG_SOC_SERIES_STM32F1X */


#define PINCTRL_SOC_PINS_ELEM_INIT(node)                \
{                                                       \
        .pinmux = DT_PROP(node, pinmux),                \
        .pincfg = Z_PINCTRL_STM32_PINCFG_INIT(node)     \
},

/**
 * @brief Utility macro to initialize each pin.
 *
 * @param node_id Node identifier.
 * @param state_prop State property name.
 * @param idx State property entry index.
 */
#define Z_PINCTRL_STATE_PIN_INIT(node_id, state_prop, idx)                     \
        PINCTRL_SOC_PINS_ELEM_INIT(DT_PROP_BY_IDX(node_id, state_prop, idx))
```

Reply:
I've simplified STM32 initializers, please check now